File: python-dev/ffbstools/noderoute.py
#!/usr/bin/env python3

import subprocess, time
from pprint import pprint

from pyroute2 import IPRoute
from pyroute2.netlink.rtnl import rt_type

from . import wireguard

TABLE = 10
RT_PROTO = 23

IDS = {
    'wg-c1': 1,
    'wg-c2': 2,
    'wg-c3': 3,
}


def get_handshake_ages():
    result = []
    now = time.time()
    wg = wireguard.get_dict()
    for iface, data in wg.items():
        peers = data['peers']
        if not len(peers)==1:
            continue
        _, peer = peers.popitem()
        result.append((now - peer['latest_handshake'], iface))
    result.sort()
    return result


def get_wg_links():
    result = {}
    ip = IPRoute()
    for link in ip.get_links():
        linkinfo = link.get_attr('IFLA_LINKINFO')
        if not linkinfo:
            continue
        info_kind = linkinfo.get_attr('IFLA_INFO_KIND')
        if not info_kind == 'wireguard':
            continue
        operstate = link.get_attr('IFLA_OPERSTATE')
        if operstate == 'DOWN':
            continue
        pprint(link)
        ifname = link.get_attr('IFLA_IFNAME')
        result[link['index']] = ifname
        result[ifname] = link['index']
    return result

def get_wg_routes():
    result = []
    ip = IPRoute()
    for route in ip.get_routes(table=TABLE):
        if not route.get_attr('RTA_DST') is None:
            continue
        if not route['dst_len'] == 0:
            continue
        if not route['type'] == rt_type['unicast']:
            continue
        if not route['proto'] == RT_PROTO:
            continue
        oif = route.get_attr('RTA_OIF')
        result.append(oif)
    return result

def set_wg_route(oif, id):
    ip = IPRoute()
    ip.route('replace',
        table=TABLE,
        proto=RT_PROTO,
        oif=oif,
        type='unicast',
        dst_len=0,
        gateway='10.0.0.{}'.format(id),
    )


def update():
    active = [iface for age, iface in get_handshake_ages() if age < 180]
    #current = get_current():
    links = get_wg_links()
    current = [links[index] for index in get_wg_routes()]
    assert len(current) <= 1
    current = current[0] if current else None

    #pprint(active)
    #pprint(current)
    if current and current in active:
        print("current route still active")
        return

    if not active:
        print("no active tunnels")
        return

    print("activating route for {}".format(active[0]))
    set_wg_route(links[active[0]], IDS[active[0]])

def main():
    update()
    time.sleep(10)

if __name__  == "__main__":
    main()

File: lua/noderoute.lua
local json = require('luci.jsonc')
local popen = require('io').popen

local TABLE = 10
local TABLE_NAME = 'freifunk'
local RT_PROTO = 'freifunk'

function check_output(cmd)
    local f = assert(io.popen(cmd, 'r'))
    local s = assert(f:read('*a'))
    f:close()
    return s
end

function sleep(n)
    os.execute("sleep " .. tonumber(n))
end

function str_split(str, pattern)
    local res = {}
    for i in string.gmatch(str, pattern) do table.insert(res, i) end
    return res
end

function tablelength(T)
    local count = 0
    for _ in pairs(T) do count = count + 1 end
    return count
end

function has_value(tab, val)
    for index, value in ipairs(tab) do
        if value == val then
            return true
        end
    end
    return false
end


function dump(foo)
    print(json.stringify(foo))
end

function get_wg_info()
    -- done
    local output = check_output('wg show all dump')
    local results = {}
    for lineRaw in string.gmatch(output, "[^\n]+") do
        local line = str_split(lineRaw, "%S+")
        if not results[line[1]] then
            local device = {}
            device['private_key'] = line[2]
            device['public_key'] = line[3]
            device['listen_port'] = tonumber(line[4])
            device['peers'] = {}
            results[line[1]] = device
        else
            local peer = {}
            if line[3] ~= "(none)" then
                peer['preshared_key'] = line[3]
            end
            peer['endpoint'] = line[4]
            peer['allowed-ips'] = str_split(line[5], "[^,]+")
            peer['latest_handshake'] = tonumber(line[6])
            peer['transfer_rx'] = tonumber(line[7])
            peer['transfer_tx'] = tonumber(line[8])
            peer['presistent_keepalive'] = tonumber(line[8])
            results[line[1]]['peers'][line[2]] = peer
        end
    end
    return results
end

function get_handshake_ages()
    -- done
    local result = {}
    local now = os.time()
    local wg = get_wg_info()
    for iface, data in pairs(wg) do
        local peers = data['peers']
        if tablelength(peers) == 1 then
            for k,v in pairs(peers) do
                table.insert(result, {now - v['latest_handshake'], iface})
            end
        end
    end
    return result
end

function get_wg_links()
    -- done, not needed
    local output = check_output('ip a')
    local result = {}
    for i_raw in string.gmatch(output, "[0-9]+: wg-[^\n]+UP") do
        local sp = str_split(i_raw, '[^: ]+')
        local i = tonumber(sp[1])
        local name = sp[2]
        result[i] = name
        result[name] = i
    end
    return result
end

function get_wg_routes()
    -- done, test needed
    local result = {}
    local output = check_output('ip r show table '..TABLE)
    for line in string.gmatch(output, "[^\n]+") do
        if string.find(line, 'default via') then
            if string.find(line, 'proto '..RT_PROTO) then
                if not string.find(line, 'broadcast') then
                    for iface in string.gmatch(line, 'dev [a-z0-9-_]+') do
                        table.insert(result, iface:sub(5))
                    end
                end
            end
        end
    end
    return result
end

function set_wg_route(iface, id)
    -- done, test needed
    os.execute('ip r replace default via 10.0.0.'..id..' dev '..iface..' proto '..RT_PROTO..' table freifunk')
end

function update()
    -- done
    local active = {}
    for _, elem in ipairs(get_handshake_ages()) do
        if elem[1] < 180 then
            table.insert(active, elem[2])
        end
    end
    local current = get_wg_routes()
    assert(#current < 1, 'too many current routes')
    current = current[1]
    if current and has_value(active, current) then
        print('current route still active')
        return
    end
    if #active == 0 then
        print('no active tunnels')
        return
    end
    print('activating route for '..active[1])
    local id = tonumber(active[1]:match('[0-9]+'))
    set_wg_route(active[1], id)
end

function main()
    -- done
    while true do
        update()
        sleep(5)
    end
end

main()




